### PR TITLE
Add remove_node_drop_children

### PR DESCRIPTION
**Pull request by Drakulix**
Todos:

Do you think there is an obvious more performant way to implement this?
Naming?

### CODE REVIEW

**Comment by iwburns** [comment]
I can't _think_ of any more performant way to implement it, but that certainly doesn't mean there's not a faster way to do it.  I hate that there have to be two `clone`s in there, but I understand why they're there.

As for naming, the only other thing I can think of is `remove_node_drop_children`.  I like it a _little bit_ better than `remove_node_remove_children`, but not by much.

**Comment by Drakulix** [comment]
I actually like `drop_children` much better.

**Comment by Drakulix** [comment]
I can get rid of both `clone`s with `take_children` from #5 
Once you merge that, I will finish this.

**Comment by iwburns** [comment]
@Drakulix #5 is merged, not rushing you, just pinging.

**Comment by Drakulix** [comment]
Done

File: src/tree.rs [hunk 1]
```diff
@@ -388,6 +388,51 @@ impl<T> Tree<T> {
...
+    /// let child_id = tree.insert_with_parent(child_node, &root_id).ok().unwrap();
+    /// tree.insert_with_parent(grandchild_node, &root_id);
+    ///
+    /// let root_node = tree.remove_node_drop_children(child_id);
+    /// ```
+    ///
+    pub fn remove_node_drop_children(&mut self, node_id: NodeId) -> Result<Node<T>, NodeIdError> {
+        let (is_valid, error) = self.is_valid_node_id(&node_id);
+        if !is_valid {
+            return Result::Err(error.expect("Tree::remove_node_orphan_children: Missing an error value on finding an invalid NodeId."));
```
Comment (iwburns) [inline src/tree.rs]: Could you fix the `expect()` string here to say "Tree::remove_node_drop_children: ..."?

File: src/tree.rs [hunk 2]
```diff
@@ -388,6 +388,51 @@ impl<T> Tree<T> {
...
+    ///
+    /// let root_node = Node::new(1);
+    /// let child_node = Node::new(2);
+    /// let grandchild_node = Node::new(3);
+    ///
+    /// let mut tree: Tree<i32> = Tree::new();
+    /// let root_id = tree.set_root(root_node);
+    ///
+    /// let child_id = tree.insert_with_parent(child_node, &root_id).ok().unwrap();
+    /// tree.insert_with_parent(grandchild_node, &root_id);
```
Comment (iwburns) [inline src/tree.rs]: Did you mean `tree.insert_with_parent(grandchild_node, &child_id);` here?

File: src/tree.rs [hunk 3]
```diff
@@ -388,6 +388,51 @@ impl<T> Tree<T> {
...
+    ///
+    /// let root_node = tree.remove_node_drop_children(child_id);
```
Comment (iwburns) [inline src/tree.rs]: Should this be `root_id` instead of `child_id`, or possibly `child_node` instead of `root_node`?

**Comment by Drakulix** [comment]
Tried to address all your criticism.
Tell me if there is anything missing!

**Comment by iwburns** [comment]
Everything looks good, but I do have one more concern. (Sorry!)

I just realized that this method is different from the other `remove_node_[something]` methods in that the existing methods don't remove other `Node`s from the `Tree` (and this one does).  This seems obvious, of course, but it does have a not-so-obvious consequence: Because child `Node`s are removed, the `NodeId`s that refer to them in the calling context will be invalid after the call is made.

I don't think this is a reason not to include the method (we return `Option`s/`Result`s for a reason), but I do think it merits some kind of mention in the documentation.  Could you add something like that in the documentation for this method?

Basically just something like: "Please keep in mind: Children of this `NodeId` _are removed during this method call_, so `NodeId`s that previously pointed to them will no longer be valid after calling this method".

**Comment by Drakulix** [comment]
Done

**Comment by iwburns** [comment]
Awesome!  Thanks, and I like that wording!  Merging now and I'll let you know when it's up on crates.io.

**Comment by iwburns** [comment]
@Drakulix it's up on crates.io
